### DISCUSSION TITLE
coupling openfast with MATLAB

### DISCUSSION THREAD

**Discussion by relhamoud** [OpenFAST/openfast#2012] [Q&A] [answered]
Hello everyone,

I would ask about coupling OpenFAST with MATLAB. The purpose of this coupling is to use the optimization toolbox buit in matlab to optimize TMD (Tuned Mass Damper) parameters.

Does anyone succeed to make this coupling ? Or is it possible to do it  ?

Thank you in advance.

Best Regards,

Riad

**Comment by jjonkman** [comment]
Dear @relhamoud,

I'm not fully sure I understand your question, but in general, I would say that OpenFAST is already coupled to MATLAB.

OpenFAST supports a direct interface to MATLAB/Simulink so that the OpenFAST dynamic library can be called from Simulink through an S-Function: https://github.com/OpenFAST/openfast/tree/main/glue-codes/simulink.

The executable of OpenFAST can also be called through the MATLAB workspace, e.g., see: https://github.com/OpenFAST/matlab-toolbox/blob/main/Utilities/runFAST.m.

Best regards,

**Reply by relhamoud** [reply]
Dear @jjonkman,

Thank you for your reply.

In fact, I am looking on how to tell MATLAB that i want to set the objective function to let's say SD of platform pitch and the most important thing is to be able to change the TMD parameters at each time step. To my knowledge, if the simulation was running and in the middle of the simulation let's say i changed the parameters of the TMD, the change will not be considered. Right ?

Concerning the interface MATLAB/Simulink, it is used mainly for control and it is limited for blade pitch control by setting **PCMODE=4** and also for VS control by setting **VS control to 4** (I think VS means Variable Speed, correct me please if i am wrong).

For MATLAB toolbox, it is generally used for post-processing the data although it contains some other features.

Thank you in advance.

Best Regards,

**Reply by jjonkman** [reply, marked as answer]
Dear @relhamoud

Regarding the TMD, it is not possible to dynamically change the parameters of the TMD (such as mass or stiffness) without changes to the source code.  I would expect if you are optimizing the TMD, that you'd run separate simulations with different TMD parameters in the optimization loop.

I generally agree with your other comments.  I would add that the MATLAB/Simulink interface to OpenFAST supports active control through Simulink of blade pitch, generator torque, nacelle-yaw, and the high-speed shaft brake.  The MATLAB/Simulink interface to OpenFAST does not currently support active control of TMDs.

FYI: Considering your interest in optimization, I would suggest considering using NREL's WEIS software: https://github.com/WISDEM/WEIS, which is targeted at system-level optimization, including of TMDs, and wraps around OpenFAST (among other NREL software).  Please note, though, that WEIS is Python based rather than MATLAB based.

Best regards,
